Source/Dafny/Resolver.cs

```diff
@@ -14628,49 +14630,110 @@ class LazyString_OnTypeEquals {
       return GetThisType(tok, (TopLevelDeclWithMembers)member.EnclosingClass);
     }
 
-    public class ResolveOpts {
-      public readonly ICodeContext codeContext;
-      public readonly bool twoState;
-      public readonly bool isReveal;
-      public readonly bool isPostCondition;
-      public readonly bool InsideOld;
+    public class ResolutionContext {
+      public readonly ICodeContext CodeContext;
+      public readonly bool IsTwoState;
+      public readonly bool InOld; // implies !IsTwoState
+      public readonly bool InReveal;
+      public readonly bool InFunctionPostcondition;
+      public readonly bool InFirstPhaseConstructor; // implies codeContext is Constructor
 
-      public ResolveOpts(ICodeContext codeContext, bool twoState) {
+      public bool IsGhost => CodeContext.IsGhost;
+
+      private ResolutionContext(ICodeContext codeContext, bool isTwoState, bool inOld, bool inReveal, bool inFunctionPostcondition,
+        bool inFirstPhaseConstructor) {
         Contract.Requires(codeContext != null);
-        this.codeContext = codeContext;
-        this.twoState = twoState;
-        this.isReveal = false;
-        this.isPostCondition = false;
+        this.CodeContext = codeContext;
+        this.IsTwoState = isTwoState;
+        this.InOld = inOld;
+        this.InReveal = inReveal;
+        this.InFunctionPostcondition = inFunctionPostcondition;
+        this.InFirstPhaseConstructor = inFirstPhaseConstructor;
+        Contract.Assert(!InOld || !IsTwoState); // InOld ==> !IsTwoState
+        Contract.Assert(!InFirstPhaseConstructor || CodeContext is Constructor);
       }
 
-      public ResolveOpts(ICodeContext codeContext, bool twoState, bool isReveal, bool isPostCondition, bool insideOld) {
-        Contract.Requires(codeContext != null);
-        this.codeContext = codeContext;
-        this.twoState = twoState;
-        this.isReveal = isReveal;
-        this.isPostCondition = isPostCondition;
-        this.InsideOld = insideOld;
+      public ResolutionContext(ICodeContext codeContext, bool isTwoState)
+        : this(codeContext, isTwoState, false, false, false, false) {
+      }
+
+      /// <summary>
+      /// Return a ResolutionContext appropriate for the body of "codeContext".
+      /// </summary>
+      public static ResolutionContext FromTopLevel(ICodeContext codeContext) {
```

Comment:
The inconsistency in the names, naming the method From _TopLevel_ followed by the type `ICodeContext`, and the comment "codeContext", make it seem you don't like the name `ICodeContext`. Please follow-up with a rename.